common/persistence/visibility_hybrid_manager.go

```diff
@@ -0,0 +1,580 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	visibilityHybridManager struct {
+		logger                    log.Logger
+		visibilityMgrs            map[string]VisibilityManager
+		readVisibilityStoreName   dynamicconfig.StringPropertyFnWithDomainFilter
+		writeVisibilityStoreName  dynamicconfig.StringPropertyFn
+		logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter
+	}
+)
+
+const (
+	ContextKey           = ResponseComparatorContextKey("visibility-override")
+	dbVisStoreName       = "db"
+	advancedWriteModeOff = "off"
+)
+
+// ResponseComparatorContextKey is for Pinot/ES response comparator. This struct will be passed into ctx as a key.
+type ResponseComparatorContextKey string
+
+type OperationType string
+
+var Operation = struct {
+	LIST  OperationType
+	COUNT OperationType
+}{
+	LIST:  "list",
+	COUNT: "count",
+}
+
+var _ VisibilityManager = (*visibilityHybridManager)(nil)
+
+// NewVisibilityTripleManager create a visibility manager that operate on DB or advanced visibility based on dynamic config.
+// For Pinot migration, Pinot is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+// For OpenSearch migration, OS is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+func NewVisibilityHybridManager(
+	visibilityMgrs map[string]VisibilityManager,
+	readVisibilityStoreName dynamicconfig.StringPropertyFnWithDomainFilter,
+	writeVisibilityStoreName dynamicconfig.StringPropertyFn,
+	logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter,
+	logger log.Logger,
+) VisibilityManager {
+	if len(visibilityMgrs) == 0 {
+		logger.Fatal("No visibility managers provided. At least one visibility manager is required.")
+		return nil
+	}
+
+	if logCustomerQueryParameter == nil {
+		logCustomerQueryParameter = dynamicconfig.GetBoolPropertyFnFilteredByDomain(false)
+	}
+
+	return &visibilityHybridManager{
+		visibilityMgrs:            visibilityMgrs,
+		readVisibilityStoreName:   readVisibilityStoreName,
+		writeVisibilityStoreName:  writeVisibilityStoreName,
+		logger:                    logger,
+		logCustomerQueryParameter: logCustomerQueryParameter,
+	}
+}
+
+func (v *visibilityHybridManager) Close() {
+	for _, mgr := range v.visibilityMgrs {
+		if mgr != nil {
+			mgr.Close()
+		}
+	}
+}
+
+func (v *visibilityHybridManager) GetName() string {
+	storeNames := strings.Split(v.writeVisibilityStoreName(), ",")
+	if mgr, ok := v.visibilityMgrs[storeNames[0]]; ok && mgr != nil {
```

Comment:
Do we need to check nil for storeNames[] ?

Reply:
We have default values for writeVisibilityStoreName, if it is updated to "", storeNames[0] returns "" and won't find a mgr for it, eventually return ""